File: lib/models/tiering/tier_coins_data.dart
import 'package:snabbit_runner/utils/common_methods.dart';
import 'package:snabbit_runner/utils/enums.dart';

/// Root payload for the tier-coins progress card (`TierNudgeProgressCard`).
///
/// Mirrors the API shape:
/// ```json
/// {
///   "coin_balance": 1250,
///   "coins": { "current_week": 3, "weeks": [ ... ] }
/// }
/// ```
class TierCoinsData {
  final int? coinBalance;
  final TierCoinsProgress? coins;

  TierCoinsData({
    this.coinBalance,
    this.coins,
  });

  factory TierCoinsData.fromJson(Map<String, dynamic> json) {
    return TierCoinsData(
      coinBalance: anyValueToInt(json['coin_balance']),
      coins: json['coins'] is Map<String, dynamic>
          ? TierCoinsProgress.fromJson(json['coins'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// Weekly coin-earning progress (the `coins` object).
class TierCoinsProgress {
  final int? currentWeek;
  final List<TierWeek>? weeks;

  TierCoinsProgress({
    this.currentWeek,
    this.weeks,
  });

  factory TierCoinsProgress.fromJson(Map<String, dynamic> json) {
    return TierCoinsProgress(
      currentWeek: anyValueToInt(json['current_week']),
      weeks: json['weeks'] is List
          ? (json['weeks'] as List)
              .whereType<Map<String, dynamic>>()
              .map(TierWeek.fromJson)
              .toList()
          : null,
    );
  }

  /// The [TierWeek] whose [TierWeek.week] equals [currentWeek], or null when
  /// there is no match. Drives the card's active-week progress + milestones.
  TierWeek? get activeWeek {
    final week = currentWeek;
    final list = weeks;
    if (week == null || list == null) return null;
    for (final w in list) {
      if (w.week == week) return w;
    }
    return null;
  }
}

/// A single week's earnings plus, for the in-progress week, its milestone
/// [targets].
class TierWeek {
  final int? week;
  final int? earned;
  final TierWeekState? state;
  final List<TierTarget>? targets;

  TierWeek({
    this.week,
    this.earned,
    this.state,
    this.targets,
  });

  factory TierWeek.fromJson(Map<String, dynamic> json) {
    return TierWeek(
      week: anyValueToInt(json['week']),
      earned: anyValueToInt(json['earned']),
      state: TierWeekState.fromString(json['state']?.toString()),
      targets: json['targets'] is List
          ? (json['targets'] as List)
              .whereType<Map<String, dynamic>>()
              .map(TierTarget.fromJson)
              .toList()
          : null,
    );
  }
}

/// A milestone on the progress bar: earn [amount] coins to reach [tier].
class TierTarget {
  final int? amount;
  final Tier? tier;

  TierTarget({
    this.amount,
    this.tier,
  });

  factory TierTarget.fromJson(Map<String, dynamic> json) {
    return TierTarget(
      amount: anyValueToInt(json['amount']),
      tier: Tier.fromString(json['tier']?.toString()),
    );
  }
}

/// Lifecycle of a [TierWeek] as reported by the API (`state`).
enum TierWeekState {
  completed,
  inProgress;

  static TierWeekState? fromString(String? value) {
    switch (value?.toLowerCase()) {
      case 'completed':
        return TierWeekState.completed;
      case 'in_progress':
        return TierWeekState.inProgress;
      default:
        return null;
    }
  }
}

File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/features/tiering/domain/model/TierCoinsData.kt
package com.snabbit.runner.shared.features.tiering.domain.model

import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.persistentListOf

/**
 * Weekly Snabbit-coins progress for the tier nudge (`THE_COIN_NUDGE`).
 *
 * Pure domain model — mirrors the Flutter `TierCoinsData` (rich weeks/targets
 * shape). No serialization here: the components render it, and the eventual
 * data layer maps its DTO into this.
 */
data class TierCoinsData(
    val coinBalance: Int? = null,
    val coins: TierCoinsProgress? = null,
)

/** The `coins` object — the weekly rollup plus which week is active. */
data class TierCoinsProgress(
    val currentWeek: Int? = null,
    val weeks: ImmutableList<TierWeek> = persistentListOf(),
) {
    /** The week whose [TierWeek.week] matches [currentWeek] (drives the card), or null. */
    val activeWeek: TierWeek?
        get() = currentWeek?.let { week -> weeks.firstOrNull { it.week == week } }
}

/** One week's earnings and, for the in-progress week, its milestone [targets]. */
data class TierWeek(
    val week: Int? = null,
    val earned: Int? = null,
    val state: TierWeekState? = null,
    val targets: ImmutableList<TierTarget> = persistentListOf(),
) {
    /** Largest milestone amount — the full scale of the progress bar (0 when none). */
    val maxTargetAmount: Int
        get() = targets.maxOfOrNull { it.amount ?: 0 } ?: 0

    /** [earned] / [maxTargetAmount], clamped to 0..1 (0 when there is no positive target). */
    val progressFraction: Float
        get() {
            val max = maxTargetAmount
            if (max <= 0) return 0f
            return ((earned ?: 0).toFloat() / max).coerceIn(0f, 1f)
        }
}

/** A milestone on the progress bar: earn [amount] coins to reach [tier]. */
data class TierTarget(
    val amount: Int? = null,
    val tier: Tier? = null,
)

/** Lifecycle of a [TierWeek] (`state`). */
enum class TierWeekState {
    COMPLETED,
    IN_PROGRESS,
    ;

    companion object {
        /** Parse the wire `state` string (case-insensitive); unknown/blank → null. */
        fun fromWire(value: String?): TierWeekState? = when (value?.trim()?.lowercase()) {
            "completed" -> COMPLETED
            "in_progress" -> IN_PROGRESS
            else -> null
        }
    }
}

/**
 * Runner tiers. [displayName] backs the header label ("Pink Diamond Tier").
 *
 * Two schemes coexist: the **legacy** (sunset) tiers [BASIC] / [PRO] / [ELITE]
 * ([isLegacyTier]) and the **new** scheme [BASE] / [SILVER] / [GOLD] / [DIAMOND] /
 * [PINK_DIAMOND]. `BASIC` is a **distinct** legacy tier — NOT an alias of the new
 * `BASE` — so a just-promoted runner (new `tier`, effective date still in the
 * future) is classified by tier identity. Mirrors Flutter's `Tier` enum.
 */
enum class Tier(val displayName: String) {
    BASIC("Basic"),
    BASE("Base"),
    PRO("Pro"),
    ELITE("Elite"),
    SILVER("Silver"),
    GOLD("Gold"),
    DIAMOND("Diamond"),
    PINK_DIAMOND("Pink Diamond"),
    ;

    /**
     * The sunset tiers (`BASIC` / `PRO` / `ELITE`) — the old scheme. The tiering
     * surfaces gate on this (tier identity), not the effective date, so a runner
     * promoted into the new scheme sees consistent chrome immediately. Mirrors
     * Flutter `Tier.isLegacyTier`.
     */
    val isLegacyTier: Boolean get() = this == BASIC || this == PRO || this == ELITE

    companion object {
        /** Parse a wire tier string (case-insensitive); unknown/blank → null (mirrors Flutter `Tier.fromString`). */
        fun fromWire(value: String?): Tier? = when (value?.trim()?.uppercase()) {
            // BASIC is a DISTINCT legacy tier — NOT an alias of the new BASE.
            "BASIC" -> BASIC
            "BASE" -> BASE
            "PRO" -> PRO
            "ELITE" -> ELITE
            "SILVER" -> SILVER
            "GOLD" -> GOLD
            "DIAMOND" -> DIAMOND
            "PINK_DIAMOND" -> PINK_DIAMOND
            else -> null
        }
    }
}
